Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch

```diff
@@ -0,0 +1,244 @@
+-                                        SpawnerCreature.a(enumcreaturetype, (World) this.world, chunk, blockposition);
++                                    // Paper start - only allow spawns upto the limit per chunk and update count afterwards
++                                    int currEntityCount = object2intmap.getInt(enumcreaturetype);
++                                    int difference = k1 - currEntityCount;
++                                    if(this.world.paperConfig.perPlayerMobSpawns){
++                                        final int[] min = {Integer.MAX_VALUE};
++                                        final int finalLimit = limit;
++                                        this.getPlayers(chunk.getPos(), this.world.spigotConfig.mobSpawnRange)
++                                            .forEach((entityplayer) -> min[0] = Math.min(
++                                                finalLimit - ((WorldServer)chunk.getWorld()).getEntityCount(entityplayer, enumcreaturetype),
++                                                min[0]));
++                                        difference = (min[0] == Integer.MAX_VALUE) ? 0 : min[0];
++                                    }
++                                    if (difference > 0) {
++                                        List spawned = SpawnerCreature.spawnMobs(enumcreaturetype, this.world, chunk, blockposition, difference);
++                                        object2intmap.put(enumcreaturetype, currEntityCount + spawned.size());
++                                        this.world.updatePlayerMobTypeMap(spawned);
++                                    // Paper end
+                                     }
+                                 }
+                             }
+@@ -441,6 +457,12 @@ public class ChunkProviderServer extends IChunkProvider {
+         this.playerChunkMap.g();
+     }
+ 
++    // Paper start
++    public Stream<EntityPlayer> getPlayers(ChunkCoordIntPair coordIntPair, int rangeChunks){
++        return this.playerChunkMap.getPlayers(coordIntPair, rangeChunks);
++    }
++    // Paper end
++
+     @Override
+     public String getName() {
+         return "ServerChunkCache: " + this.g();
+diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+index 86831c35..e533cce0 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
++++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+@@ -974,6 +974,14 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+         });
+     }
+ 
++    // Paper start
++    public Stream<EntityPlayer> getPlayers(ChunkCoordIntPair chunkcoordintpair, int range) {
++        int blockRange = ((range << 4) * (range << 4));
++        return this.playerMap.a(chunkcoordintpair.pair()).filter((entityplayer) ->
++            (!entityplayer.isSpectator() && a(chunkcoordintpair, (Entity) entityplayer) < blockRange));
```

Comment:
obfhelpers

Reply:
@Spottedleaf wasn't talking about comment lines here.
You need an OPFHELPER if you call obfuscated methods like a(). So here for PlayerMap.a() and PlayerChunkMap.a()

Reply:
Made some changes, please take a look.

Reply:
It looks fine, but OBFHELPERS won't trigger a conflict if there is an empty line between the OBFHELPER and the obfuscated method.